apps/bettafish/src/app/pages/tags/tag-page.component.ts

```diff
@@ -1,25 +1,73 @@
 import { Component, OnInit } from "@angular/core";
-import { ActivatedRoute } from "@angular/router";
+import { ActivatedRoute, Router } from "@angular/router";
 import { TagsQuery, TagsService } from "@dragonfish/client/repository/tags";
+import { DragonfishNetworkService } from "@dragonfish/client/services";
+import { setTwoPartTitle } from "@dragonfish/shared/constants";
+import { htmlDecode } from "@dragonfish/shared/functions";
+import { ContentModel, TagsModel } from "@dragonfish/shared/models/content";
+import { PaginateResult } from "@dragonfish/shared/models/util";
+import { UntilDestroy, untilDestroyed } from "@ngneat/until-destroy";
+import { combineLatest } from "rxjs";
 
+@UntilDestroy()
 @Component({
     selector: 'dragonfish-tag-page',
     templateUrl: './tag-page.component.html',
 })
 export class TagPageComponent implements OnInit {
     tagId = "";
     tagName = "";
+    pageNum = 1;
+    searchResults: PaginateResult<ContentModel>;
 
     constructor(
         public route: ActivatedRoute,
+        private router: Router,
         public tagsQuery: TagsQuery,
         private tagsService: TagsService,
+        private network: DragonfishNetworkService,
     ) {}
 
     ngOnInit(): void {
-        this.route.paramMap.subscribe((params) => {
-            this.tagId = params.get("tagId");
-            this.tagsService.fetchDescendants(this.tagId).subscribe();
+        combineLatest([this.route.paramMap, this.route.queryParamMap])
+            .pipe(untilDestroyed(this))
+            .subscribe((value) => {
+                const [params, queryParams] = value;
+                this.tagId = params.get("tagId");
+                if (queryParams.has('page')) {
+                    this.pageNum = +params.get('page');
+                }
+                this.fetchData(this.tagId, this.pageNum);
+            });
+    }
+
+    private fetchData(tagId: string, pageNum: number): void {
+        this.tagsService.fetchDescendants(tagId).subscribe((tagsTree) => {
+            if (tagsTree.parent) {
+                setTwoPartTitle(htmlDecode((tagsTree.parent as TagsModel).name + " — " + tagsTree.name));
+            }
+            else {
+                setTwoPartTitle(htmlDecode(tagsTree.name));
+            }
+        });
```

Comment:
We can save some network traffic here. As currently written, we're re-fetching tag data on every page change, but we can cache it, because it's unlikely to change. You'll want to call it once, then just use the cached value, i.e. something like this:

```TypeScript
import { lastValueFrom } from 'rxjs';

// As a class-level variable:
private cachedTags: TagsTree = null;

private async fetchData(tagId: string, pageNum: number): Promise<void> { // now that this is async, remember to make the callback passed to .subscribe() up above async as well)
  if (!this.cachedTags) {
    this.cachedTags = await lastValueFrom(this.tagsService.fetchDescendants(tagId));
    if (this.cachedTags.parent) {
      // etc
    }
    else {
     // etc
    } 
  }
// Though if you wanted, you could turn this into an awaitable promise with lastValueFrom() too!
this.network.searchFandomTagContent(tagId, pageNum).subscribe((results) => {
  this.searchResults = results;
});
}

```